sys/net/transport_layer/ng_udp/ng_udp.c

```diff
@@ -0,0 +1,249 @@
+
+    switch (pseudo_hdr->type) {
+#ifdef MODULE_NG_IPV6
+        case NG_NETTYPE_IPV6:
+            csum += ng_ipv6_hdr_inet_csum(csum, pseudo_hdr->data,
+                                          NG_PROTONUM_UDP, len);
+#endif
+        default:
+            return 0;
+    }
+    /* invert checksum if 0 */
+    if (csum == 0) {
+        return 0xffff;
+    }
+    return csum;
+}
+
+static void _receive(ng_pktsnip_t *pkt)
+{
+    ng_pktsnip_t *udp;
+    ng_udp_hdr_t *hdr;
+    uint16_t csum;
+    uint32_t port;
+    ng_netreg_entry_t *sendto;
+
+    /* mark UDP header */
+    pkt = ng_pktbuf_start_write(pkt);
+    if (pkt == NULL) {
+        DEBUG("udp: unable to get write access to packet\n");
+        return;
+    }
+    udp = ng_pktbuf_add(pkt, pkt->data, sizeof(ng_udp_hdr_t), NG_NETTYPE_UDP);
+    if (udp == NULL) {
+        DEBUG("udp: error marking UDP header, dropping packet\n");
+        ng_pktbuf_release(pkt);
+        return;
+    }
+    hdr = (ng_udp_hdr_t *)udp->data;
+
+    /* validate checksum */
+    csum = _calc_csum(udp, udp->next, pkt);     /* TODO: what about extension headers? */
+    if (byteorder_ntohs(hdr->checksum) != csum) {
+        DEBUG("udp: received packet with invalid checksum, dropping it\n");
+        ng_pktbuf_release(pkt);
+        return;
+    }
+
+    /* get port (netreg demux context) */
+    port = (uint32_t)byteorder_ntohs(hdr->dst_port);
+
+    /* send payload to receivers */
+    sendto = ng_netreg_lookup(NG_NETTYPE_UDP, port);
+    if (sendto == NULL) {
+        DEBUG("udp: unable to forward packet as no one is interested in it\n");
+        ng_pktbuf_release(pkt);
+        return;
+    }
+    ng_pktbuf_hold(pkt, ng_netreg_num(NG_NETTYPE_UDP, port) - 1);
+    while (sendto != NULL) {
+        ng_netapi_receive(sendto->pid, pkt);
+        sendto = ng_netreg_getnext(sendto);
+    }
+}
+
+static void _send(ng_pktsnip_t *pkt)
+{
+    ng_udp_hdr_t *hdr;
+    ng_pktsnip_t *udp_snip;
+    ng_netreg_entry_t *sendto;
+
+    /* get udp snip and hdr */
+    LL_SEARCH_SCALAR(pkt, udp_snip, type, NG_NETTYPE_UDP);
+    ng_pktbuf_start_write(udp_snip);
+    hdr = (ng_udp_hdr_t *)udp_snip->data;
+    /* fill in size field */
+    hdr->length = byteorder_htons(pkt->next->next->size + sizeof(ng_udp_hdr_t));
+
+    /* and forward packet to the network layer */
+    sendto = ng_netreg_lookup(pkt->type, NG_NETREG_DEMUX_CTX_ALL);
+    /* throw away packet if no one is interested */
+    if (sendto == NULL) {
+        DEBUG("udp: cannot send packet because network layer not found\n");
+        ng_pktbuf_release(pkt);
+        return;
+    }
+    /* send packet to network layer */
+    ng_pktbuf_hold(pkt, ng_netreg_num(pkt->type, NG_NETREG_DEMUX_CTX_ALL) - 1);
+    while (sendto != NULL) {
+        ng_netapi_send(sendto->pid, pkt);
+        sendto = ng_netreg_getnext(sendto);
+    }
+}
+
+static void *_event_loop(void *arg)
+{
+    (void)arg;
+    msg_t msg;
+    msg_t msg_queue[NG_UDP_MSG_QUEUE_SIZE];
```

Comment:
`netreg` registration is missing as pointed out in https://github.com/RIOT-OS/RIOT/pull/2430#issuecomment-94015464